File: scripts/q1_model.jl
# Question 1 model construction and LP solving.

using TyOptimization
using TyMath
using TyBase

const Q1_INTERVAL_COUNT = 144
const Q1_DELTA_HOUR = 1.0 / 6.0

const Q1_BATTERY_CAPACITY_KWH = 12000.0
const Q1_SOC_MIN_KWH = 1200.0
const Q1_SOC_MAX_KWH = 10800.0
const Q1_SOC_INITIAL_KWH = 6000.0
const Q1_SOC_TERMINAL_KWH = 6000.0

const Q1_CHARGE_EFFICIENCY = 0.9
const Q1_DISCHARGE_EFFICIENCY = 0.9
const Q1_POWER_LIMIT_KW = 5000.0

function q1_energy_from_power_kw(power_kw)
    return power_kw .* Q1_DELTA_HOUR
end

function q1_charge_energy_limit_kwh()
    return Q1_POWER_LIMIT_KW * Q1_DELTA_HOUR * Q1_CHARGE_EFFICIENCY
end

function q1_discharge_internal_limit_kwh()
    return Q1_POWER_LIMIT_KW * Q1_DELTA_HOUR / Q1_DISCHARGE_EFFICIENCY
end

function q1_index_grid(t)
    return t
end

function q1_index_soc(t)
    return Q1_INTERVAL_COUNT + t
end

function q1_variable_count()
    return 2 * Q1_INTERVAL_COUNT
end

function q1_validate_input_lengths(price, load_kw, pv_kw)
    if length(price) != Q1_INTERVAL_COUNT
        error("price length must be 144")
    end
    if length(load_kw) != Q1_INTERVAL_COUNT
        error("load_kw length must be 144")
    end
    if length(pv_kw) != Q1_INTERVAL_COUNT
        error("pv_kw length must be 144")
    end
end

function q1_prepare_interval_data(price, load_kw, pv_kw)
    q1_validate_input_lengths(price, load_kw, pv_kw)

    load_kwh = q1_energy_from_power_kw(load_kw)
    pv_kwh = q1_energy_from_power_kw(pv_kw)
    net_kwh = load_kwh .- pv_kwh

    return Dict(
        "price" => price,
        "load_kw" => load_kw,
        "pv_kw" => pv_kw,
        "load_kwh" => load_kwh,
        "pv_kwh" => pv_kwh,
        "net_kwh" => net_kwh,
    )
end

function q1_objective(price)
    f = zeros(q1_variable_count())
    for t in 1:Q1_INTERVAL_COUNT
        f[q1_index_grid(t)] = price[t]
    end
    return f
end

function q1_build_lp(interval_data)
    price = interval_data["price"]
    load_kwh = interval_data["load_kwh"]
    net_kwh = interval_data["net_kwh"]
    nvars = q1_variable_count()

    # Four inequalities per interval:
    # two epigraph constraints and two internal-energy increment bounds.
    A = zeros(4 * Q1_INTERVAL_COUNT, nvars)
    b = zeros(4 * Q1_INTERVAL_COUNT)

    row = 0
    charge_upper = q1_charge_energy_limit_kwh()
    ac_discharge_limit = Q1_POWER_LIMIT_KW * Q1_DELTA_HOUR

    for t in 1:Q1_INTERVAL_COUNT
        g = q1_index_grid(t)
        s = q1_index_soc(t)
        s_prev = t == 1 ? 0 : q1_index_soc(t - 1)

        # g_t >= n_t + x_t / eta_c
        row += 1
        A[row, g] = -1.0
        A[row, s] = 1.0 / Q1_CHARGE_EFFICIENCY
        if t == 1
            b[row] = -net_kwh[t] + Q1_SOC_INITIAL_KWH / Q1_CHARGE_EFFICIENCY
        else
            A[row, s_prev] = -1.0 / Q1_CHARGE_EFFICIENCY
            b[row] = -net_kwh[t]
        end

        # g_t >= n_t + eta_d * x_t
        row += 1
        A[row, g] = -1.0
        A[row, s] = Q1_DISCHARGE_EFFICIENCY
        if t == 1
            b[row] = -net_kwh[t] + Q1_DISCHARGE_EFFICIENCY * Q1_SOC_INITIAL_KWH
        else
            A[row, s_prev] = -Q1_DISCHARGE_EFFICIENCY
            b[row] = -net_kwh[t]
        end

        # x_t <= eta_c * Pmax * dt
        row += 1
        A[row, s] = 1.0
        if t == 1
            b[row] = charge_upper + Q1_SOC_INITIAL_KWH
        else
            A[row, s_prev] = -1.0
            b[row] = charge_upper
        end

        # x_t >= -min(Pmax*dt, load_t) / eta_d
        discharge_lower = -min(ac_discharge_limit, load_kwh[t]) / Q1_DISCHARGE_EFFICIENCY
        row += 1
        A[row, s] = -1.0
        if t == 1
            b[row] = -discharge_lower - Q1_SOC_INITIAL_KWH
        else
            A[row, s_prev] = 1.0
            b[row] = -discharge_lower
        end
    end

    Aeq = zeros(1, nvars)
    Aeq[1, q1_index_soc(Q1_INTERVAL_COUNT)] = 1.0
    beq = [Q1_SOC_TERMINAL_KWH]

    lb = zeros(nvars)
    ub = fill(Inf, nvars)
    for t in 1:Q1_INTERVAL_COUNT
        s = q1_index_soc(t)
        lb[s] = Q1_SOC_MIN_KWH
        ub[s] = Q1_SOC_MAX_KWH
    end

    return Dict(
        "f" => q1_objective(price),
        "A" => A,
        "b" => b,
        "Aeq" => Aeq,
        "beq" => beq,
        "lb" => lb,
        "ub" => ub,
    )
end

function q1_recover_solution(z, interval_data; name="solution")
    grid_kwh = vec(z[1:Q1_INTERVAL_COUNT])
    soc_kwh = vec(z[(Q1_INTERVAL_COUNT + 1):end])

    previous_soc = vcat([Q1_SOC_INITIAL_KWH], soc_kwh[1:end-1])
    internal_delta = soc_kwh .- previous_soc
    charge_kwh = max.(internal_delta, 0.0) ./ Q1_CHARGE_EFFICIENCY
    discharge_kwh = Q1_DISCHARGE_EFFICIENCY .* max.(-internal_delta, 0.0)
    curtail_kwh = grid_kwh .+ interval_data["pv_kwh"] .+ discharge_kwh .-
                  interval_data["load_kwh"] .- charge_kwh

    return Dict(
        "name" => name,
        "z" => z,
        "grid_kwh" => grid_kwh,
        "soc_kwh" => soc_kwh,
        "internal_delta_kwh" => internal_delta,
        "charge_kwh" => charge_kwh,
        "discharge_kwh" => discharge_kwh,
        "curtail_kwh" => curtail_kwh,
        "cost" => sum(interval_data["price"] .* grid_kwh),
        "total_grid_kwh" => sum(grid_kwh),
        "total_charge_kwh" => sum(charge_kwh),
        "total_discharge_kwh" => sum(discharge_kwh),
        "total_curtail_kwh" => sum(curtail_kwh),
    )
end

function q1_solve_lp(lp; objective=nothing, Aeq=nothing, beq=nothing, name="solution")
    f = isnothing(objective) ? lp["f"] : objective
    local_Aeq = isnothing(Aeq) ? lp["Aeq"] : Aeq
    local_beq = isnothing(beq) ? lp["beq"] : beq

    options = optimoptions(:linprog, Display = "off")
    z, fval, exitflag, output = linprog(
        vec(f),
        lp["A"],
        lp["b"],
        local_Aeq,
        local_beq,
        vec(lp["lb"]),
        vec(lp["ub"]),
        options,
    )

    if exitflag <= 0
        error("linprog failed for " * name * ", exitflag = " * string(exitflag))
    end

    return z, fval, exitflag, output
end

function q1_solve_primary(interval_data)
    lp = q1_build_lp(interval_data)
    z, fval, exitflag, output = q1_solve_lp(lp, name="primary-cost")
    solution = q1_recover_solution(z, interval_data, name="primary-cost")
    solution["reported_fval"] = fval
    solution["exitflag"] = exitflag
    return solution, lp
end

function q1_solve_at_same_cost(interval_data, lp, optimal_cost; target_index=140, sense="min")
    nvars = q1_variable_count()
    f2 = zeros(nvars)
    f2[q1_index_grid(target_index)] = sense == "max" ? -1.0 : 1.0

    cost_row = reshape(lp["f"], 1, nvars)
    Aeq2 = vcat(lp["Aeq"], cost_row)
    beq2 = vcat(lp["beq"], [optimal_cost])

    z, fval, exitflag, output = q1_solve_lp(
        lp,
        objective=f2,
        Aeq=Aeq2,
        beq=beq2,
        name=sense * "-g" * string(target_index),
    )

    solname = sense == "max" ? "plan_B_max_g$(target_index)" : "plan_A_min_g$(target_index)"
    solution = q1_recover_solution(z, interval_data, name=solname)
    solution["reported_fval"] = fval
    solution["exitflag"] = exitflag
    solution["target_index"] = target_index
    solution["target_grid_kwh"] = solution["grid_kwh"][target_index]
    return solution
end

function q1_no_storage_baseline(interval_data)
    grid_kwh = max.(interval_data["net_kwh"], 0.0)
    curtail_kwh = max.(-interval_data["net_kwh"], 0.0)
    return Dict(
        "name" => "no_storage",
        "grid_kwh" => grid_kwh,
        "cost" => sum(interval_data["price"] .* grid_kwh),
        "total_grid_kwh" => sum(grid_kwh),
        "total_curtail_kwh" => sum(curtail_kwh),
    )
end

File: scripts/q1_model.py
"""Question 1 deterministic battery scheduling model."""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from scipy.optimize import linprog

from q1_data import DELTA_HOUR, INTERVAL_COUNT, Q1SourceData

BATTERY_CAPACITY_KWH = 12000.0
SOC_MIN_KWH = 1200.0
SOC_MAX_KWH = 10800.0
SOC_INITIAL_KWH = 6000.0
SOC_TERMINAL_KWH = 6000.0
CHARGE_EFFICIENCY = 0.9
DISCHARGE_EFFICIENCY = 0.9
POWER_LIMIT_KW = 5000.0


@dataclass(frozen=True)
class IntervalData:
    price: np.ndarray
    load_kw: np.ndarray
    pv_kw: np.ndarray
    load_kwh: np.ndarray
    pv_kwh: np.ndarray
    net_kwh: np.ndarray


@dataclass(frozen=True)
class LinearProgram:
    objective: np.ndarray
    a_ub: np.ndarray
    b_ub: np.ndarray
    a_eq: np.ndarray
    b_eq: np.ndarray
    bounds: list[tuple[float | None, float | None]]


@dataclass(frozen=True)
class ScheduleSolution:
    name: str
    z: np.ndarray
    grid_kwh: np.ndarray
    soc_kwh: np.ndarray
    internal_delta_kwh: np.ndarray
    charge_kwh: np.ndarray
    discharge_kwh: np.ndarray
    curtail_kwh: np.ndarray
    cost: float
    total_grid_kwh: float
    total_charge_kwh: float
    total_discharge_kwh: float
    total_curtail_kwh: float
    target_index: int | None = None
    target_grid_kwh: float | None = None
    solver_status: int | None = None
    solver_message: str | None = None


@dataclass(frozen=True)
class BaselineSolution:
    name: str
    grid_kwh: np.ndarray
    curtail_kwh: np.ndarray
    cost: float
    total_grid_kwh: float
    total_curtail_kwh: float


def energy_from_power_kw(power_kw: np.ndarray) -> np.ndarray:
    return np.asarray(power_kw, dtype=float) * DELTA_HOUR


def prepare_interval_data(source: Q1SourceData) -> IntervalData:
    if len(source.price_yuan_per_kwh) != INTERVAL_COUNT:
        raise ValueError("price length must be 144")
    load_kwh = energy_from_power_kw(source.load_kw)
    pv_kwh = energy_from_power_kw(source.pv_forecast_kw)
    return IntervalData(
        price=np.asarray(source.price_yuan_per_kwh, dtype=float),
        load_kw=np.asarray(source.load_kw, dtype=float),
        pv_kw=np.asarray(source.pv_forecast_kw, dtype=float),
        load_kwh=load_kwh,
        pv_kwh=pv_kwh,
        net_kwh=load_kwh - pv_kwh,
    )


def grid_index(t: int) -> int:
    return t


def soc_index(t: int) -> int:
    return INTERVAL_COUNT + t


def variable_count() -> int:
    return 2 * INTERVAL_COUNT


def charge_internal_limit_kwh() -> float:
    return POWER_LIMIT_KW * DELTA_HOUR * CHARGE_EFFICIENCY


def build_lp(interval_data: IntervalData) -> LinearProgram:
    nvars = variable_count()
    objective = np.zeros(nvars)
    objective[:INTERVAL_COUNT] = interval_data.price

    a_ub = np.zeros((4 * INTERVAL_COUNT, nvars))
    b_ub = np.zeros(4 * INTERVAL_COUNT)
    charge_upper = charge_internal_limit_kwh()
    ac_discharge_limit = POWER_LIMIT_KW * DELTA_HOUR

    row = 0
    for t in range(INTERVAL_COUNT):
        g = grid_index(t)
        s = soc_index(t)
        previous_s = soc_index(t - 1) if t > 0 else None
        net = interval_data.net_kwh[t]

        # g_t >= n_t + x_t / eta_c
        a_ub[row, g] = -1.0
        a_ub[row, s] = 1.0 / CHARGE_EFFICIENCY
        if previous_s is None:
            b_ub[row] = -net + SOC_INITIAL_KWH / CHARGE_EFFICIENCY
        else:
            a_ub[row, previous_s] = -1.0 / CHARGE_EFFICIENCY
            b_ub[row] = -net
        row += 1

        # g_t >= n_t + eta_d * x_t
        a_ub[row, g] = -1.0
        a_ub[row, s] = DISCHARGE_EFFICIENCY
        if previous_s is None:
            b_ub[row] = -net + DISCHARGE_EFFICIENCY * SOC_INITIAL_KWH
        else:
            a_ub[row, previous_s] = -DISCHARGE_EFFICIENCY
            b_ub[row] = -net
        row += 1

        # x_t <= eta_c * Pmax * dt
        a_ub[row, s] = 1.0
        if previous_s is None:
            b_ub[row] = charge_upper + SOC_INITIAL_KWH
        else:
            a_ub[row, previous_s] = -1.0
            b_ub[row] = charge_upper
        row += 1

        # x_t >= -min(Pmax*dt, load_t) / eta_d
        discharge_lower = -min(ac_discharge_limit, interval_data.load_kwh[t]) / DISCHARGE_EFFICIENCY
        a_ub[row, s] = -1.0
        if previous_s is None:
            b_ub[row] = -discharge_lower - SOC_INITIAL_KWH
        else:
            a_ub[row, previous_s] = 1.0
            b_ub[row] = -discharge_lower
        row += 1

    a_eq = np.zeros((1, nvars))
    a_eq[0, soc_index(INTERVAL_COUNT - 1)] = 1.0
    b_eq = np.array([SOC_TERMINAL_KWH])

    bounds: list[tuple[float | None, float | None]] = [(0.0, None) for _ in range(nvars)]
    for t in range(INTERVAL_COUNT):
        bounds[soc_index(t)] = (SOC_MIN_KWH, SOC_MAX_KWH)

    return LinearProgram(objective, a_ub, b_ub, a_eq, b_eq, bounds)


def solve_lp(
    lp: LinearProgram,
    *,
    objective: np.ndarray | None = None,
    a_eq: np.ndarray | None = None,
    b_eq: np.ndarray | None = None,
    name: str,
):
    result = linprog(
        c=lp.objective if objective is None else objective,
        A_ub=lp.a_ub,
        b_ub=lp.b_ub,
        A_eq=lp.a_eq if a_eq is None else a_eq,
        b_eq=lp.b_eq if b_eq is None else b_eq,
        bounds=lp.bounds,
        method="highs",
    )
    if not result.success:
        raise RuntimeError(f"linprog failed for {name}: {result.message}")
    return result


def recover_solution(z: np.ndarray, interval_data: IntervalData, *, name: str) -> ScheduleSolution:
    grid_kwh = np.asarray(z[:INTERVAL_COUNT], dtype=float)
    soc_kwh = np.asarray(z[INTERVAL_COUNT:], dtype=float)
    previous_soc = np.concatenate(([SOC_INITIAL_KWH], soc_kwh[:-1]))
    internal_delta = soc_kwh - previous_soc
    charge_kwh = np.maximum(internal_delta, 0.0) / CHARGE_EFFICIENCY
    discharge_kwh = DISCHARGE_EFFICIENCY * np.maximum(-internal_delta, 0.0)
    curtail_kwh = grid_kwh + interval_data.pv_kwh + discharge_kwh - interval_data.load_kwh - charge_kwh
    return ScheduleSolution(
        name=name,
        z=np.asarray(z, dtype=float),
        grid_kwh=grid_kwh,
        soc_kwh=soc_kwh,
        internal_delta_kwh=internal_delta,
        charge_kwh=charge_kwh,
        discharge_kwh=discharge_kwh,
        curtail_kwh=curtail_kwh,
        cost=float(np.dot(interval_data.price, grid_kwh)),
        total_grid_kwh=float(np.sum(grid_kwh)),
        total_charge_kwh=float(np.sum(charge_kwh)),
        total_discharge_kwh=float(np.sum(discharge_kwh)),
        total_curtail_kwh=float(np.sum(curtail_kwh)),
    )


def solve_primary(interval_data: IntervalData) -> tuple[ScheduleSolution, LinearProgram]:
    lp = build_lp(interval_data)
    result = solve_lp(lp, name="primary-cost")
    solution = recover_solution(result.x, interval_data, name="primary-cost")
    return solution.__class__(**{**solution.__dict__, "solver_status": result.status, "solver_message": result.message}), lp


def solve_at_same_cost(
    interval_data: IntervalData,
    lp: LinearProgram,
    optimal_cost: float,
    *,
    target_index: int = 140,
    sense: str = "min",
) -> ScheduleSolution:
    if not 1 <= target_index <= INTERVAL_COUNT:
        raise ValueError("target_index is one-based and must be between 1 and 144")

    nvars = variable_count()
    objective = np.zeros(nvars)
    target_zero_based = target_index - 1
    objective[grid_index(target_zero_based)] = -1.0 if sense == "max" else 1.0

    cost_row = lp.objective.reshape(1, nvars)
    a_eq = np.vstack([lp.a_eq, cost_row])
    b_eq = np.concatenate([lp.b_eq, [optimal_cost]])
    result = solve_lp(lp, objective=objective, a_eq=a_eq, b_eq=b_eq, name=f"{sense}-g{target_index}")
    solname = f"plan_B_max_g{target_index}" if sense == "max" else f"plan_A_min_g{target_index}"
    solution = recover_solution(result.x, interval_data, name=solname)
    return solution.__class__(
        **{
            **solution.__dict__,
            "target_index": target_index,
            "target_grid_kwh": float(solution.grid_kwh[target_zero_based]),
            "solver_status": result.status,
            "solver_message": result.message,
        }
    )


def no_storage_baseline(interval_data: IntervalData) -> BaselineSolution:
    grid_kwh = np.maximum(interval_data.net_kwh, 0.0)
    curtail_kwh = np.maximum(-interval_data.net_kwh, 0.0)
    return BaselineSolution(
        name="no_storage",
        grid_kwh=grid_kwh,
        curtail_kwh=curtail_kwh,
        cost=float(np.dot(interval_data.price, grid_kwh)),
        total_grid_kwh=float(np.sum(grid_kwh)),
        total_curtail_kwh=float(np.sum(curtail_kwh)),
    )
